apps/librepcb-cli/commandlineinterface.cpp

```diff
@@ -0,0 +1,334 @@
+                foreach (const QString& boardName, boards) {
+                    Board* board = project.getBoardByName(boardName);
+                    if (board) {
+                        boardList.append(board);
+                    } else {
+                        printErr(QString(tr("ERROR: No board with the name '%1' found."))
+                                 .arg(boardName));
+                        addToExitCode(exitCode, 1); // the application must report failure!
+                    }
+                }
+            }
+            QHash<FilePath, int> filesCounter;
+            bool filesOverwritten = false;
+            foreach (const Board* board, boardList) {
+                print("  " % QString(tr("Board '%1':")).arg(*board->getName()));
+                BoardGerberExport grbExport(*board);
+                grbExport.exportAllLayers(); // can throw
+                foreach (const FilePath& fp, grbExport.getWrittenFiles()) {
+                    filesCounter[fp]++;
+                    if (filesCounter[fp] > 1) filesOverwritten = true;
+                    print(QString("    => '%1'").arg(prettyPath(fp, projectFile)));
+                }
+            }
+            if (filesOverwritten) {
+                printErr("  " % tr("ERROR: Some files were written multiple times! "
+                                   "Please make sure that every board uses a different "
+                                   "fabrication output path or specify the board to "
+                                   "export with the '--board' argument."));
+                addToExitCode(exitCode, 1); // the application must report failure!
+            }
+            print(doneMsg);
+        }
+
+        // Save project
+        if (save) {
+            print(tr("Save project..."));
+            // first save to temporary files, then to original files
+            project.save(false); // can throw
+            project.save(true); // can throw
+            print(doneMsg);
+        }
+
+        return exitCode;
+    } catch (const Exception& e) {
+        printErr(QString(tr("ERROR: %1")).arg(e.getMsg()));
+        return ExitCode::COMMAND_EXECUTION_ERROR;
+    }
+}
+
+QString CommandLineInterface::prettyPath(const FilePath& path, const QString& style) noexcept
+{
+    return QFileInfo(style).isRelative()
+            ? path.toRelative(FilePath(QDir::currentPath()))
+            : path.toStr();
+}
+
+void CommandLineInterface::print(const QString& str, int newlines) noexcept
+{
+    QTextStream s(stdout);
+    s << str;
+    for (int i = 0; i < newlines; ++i) {
+        s << endl;
+    }
+}
+
+void CommandLineInterface::printErr(const QString& str, int newlines) noexcept
+{
+    QTextStream s(stderr);
+    s << str;
+    for (int i = 0; i < newlines; ++i) {
+        s << endl;
+    }
+}
+
+void CommandLineInterface::addToExitCode(qint16& exitCode, int count) noexcept
+{
+    exitCode = qMin(static_cast<int>(exitCode) + count, INT16_MAX);
+}
```

Comment:
I'm not sure if using the exit code to convey data is a good idea. What if we would use the exit code like a bitmask? The LSB bit could indicate ERC errors, for example. Another bit could indicate files-overwritten. The error count could be printed to stderr in a well-defined format instead. That would give us more flexibility regarding the meaning of the exit code.

Reply:
Hmm yeah now I also read some advices to not use exit codes other than 0 (SUCCESS) and 1 (FAILURE). Or at least only 0-127 for our own error codes. See for example https://stackoverflow.com/questions/18248511/program-exit-status-conventions

I think I'll change it to only return 0 or 1, the error reason is printed to stderr anyway...

Reply:
I like being able to determine the type of error from the status code. But I'd try to stick within the range -128 to 127.

Reply:
What would be the use-case of that? If you want to know if the ERC failed or the Gerber export failed, just perform these tasks as two separate commands instead of one combined command:

    librepcb-cli open-project --check-erc MyProject.lpp
    librepcb-cli open-project --export-gerber MyProject.lpp

Reply:
I don't have a concrete use case, just in general I think it's nice if the exit code says something about the nature of the failure :)